libfwupd/fwupd-client.c

```diff
@@ -469,6 +469,65 @@ fwupd_client_get_device_by_id (FwupdClient *client,
 	return NULL;
 }
 
+/**
+ * fwupd_client_get_device_by_guid:
+ * @client: A #FwupdClient
+ * @guid: the GUID, e.g. `e22c4520-43dc-5bb3-8245-5787fead9b63`
+ * @cancellable: the #GCancellable, or %NULL
+ * @error: the #GError, or %NULL
+ *
+ * Gets a device by its provided GUID. An error is returned if more than one
+ * active device contains this GUID.
+ *
+ * Returns: (transfer full): a #FwupdDevice or %NULL
+ *
+ * Since: 1.4.1
+ **/
+FwupdDevice *
+fwupd_client_get_device_by_guid (FwupdClient *client,
+				 const gchar *guid,
+				 GCancellable *cancellable,
+				 GError **error)
+{
+	g_autoptr(GPtrArray) devices = NULL;
+	g_autoptr(FwupdDevice) dev = NULL;
+
+	g_return_val_if_fail (FWUPD_IS_CLIENT (client), NULL);
+	g_return_val_if_fail (guid != NULL, NULL);
+	g_return_val_if_fail (cancellable == NULL || G_IS_CANCELLABLE (cancellable), NULL);
+	g_return_val_if_fail (error == NULL || *error == NULL, NULL);
+
+	/* get all the devices */
+	devices = fwupd_client_get_devices (client, cancellable, error);
+	if (devices == NULL)
+		return NULL;
+
+	/* find the device by GUID (client side) */
+	for (guint i = 0; i < devices->len; i++) {
+		FwupdDevice *dev_tmp = g_ptr_array_index (devices, i);
+		if (fwupd_device_has_guid (dev_tmp, guid)) {
+			if (dev == NULL) {
+				dev = g_object_ref (dev_tmp);
+				continue;
+			}
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "multiple devices provided %s", guid);
+			return NULL;
```

Comment:
I think specifically I would be happier with the client library returning an array of matched devices.  Then the command line frontends can use this for the prompting list, or adopt a behavior that is "update all" or similar.

Reply:
Good idea.